### PR TITLE
Container and user interface implementation

### PR DESCRIPTION
**Pull request by Zavioer**
Container class handling set up for simulation and logic. Attempt to create abstract class for Container add few unitest.

### CODE REVIEW

File: simulation/container_interface.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        self.time_to_live = time_to_live
+        self.action_interval = action_interval
+        self.move_distance_length = move_dist_length
+        self.object_list = []
+        self.imember = ''
+        self.smember = ''
+        self.rmember = ''
+        self.dmember = ''
+
+    def count_objects_in_list(self, class_definition: 'classtype') -> int:
```
Comment (amasend) [inline simulation/container_interface.py]: Please always check errors in lintern: https://github.com/amasend/SIR-simulation-IBM-ESI/pull/31/checks?check_run_id=732534648

File: simulation/container_interface.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        self.height = size
+        self.population = population
+        self.time_to_live = time_to_live
+        self.action_interval = action_interval
+        self.move_distance_length = move_dist_length
+        self.object_list = []
+        self.imember = ''
+        self.smember = ''
+        self.rmember = ''
+        self.dmember = ''
```
Comment (amasend) [inline simulation/container_interface.py]: Why these are initialized as strings?
If something should be initialized not at the beginning, the common way is to make it as `None`.

File: simulation/container_interface.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        Method checks if container instance time_to_live attribute is
+        greater than 0.
+        """
+
+        if self.time_to_live > 0:
+            return True
+        else:
+            return False
+
+    def set_population_groups(self, infection_prob: float, infection_range: float,
```
Comment (amasend) [inline simulation/container_interface.py]: What is the purpose of that method?

File: simulation/container.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        Parameters
+        ----------
+        susceptible: int, required
+            Amount of susceptible instances to place in container.
+        infected: int, required
+            Amount of infected instances to place in container.
+        recovered: int, required
+            Amount of infected instances to place in container.
+        dead: int, required
+            Amount of dead instances to place in container.
```
Comment (amasend) [inline simulation/container.py:115]: Names are confusing, maybe something like `number_of_dead` etc will be better?

File: simulation/container.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                instance.move(x, y, self.width, self.height)
+
+                if isinstance(instance, groups.SGroup):
+                    for infected in self.object_list:
+                        if isinstance(infected, groups.IGroup):
+                            if instance.is_in_infection_area(infected.x, infected.y,
+                                                             infected.infection_range):
+                                if instance.infect():
+                                    self.object_list.remove(instance)
+                                    self.add_instances(self.imember, 1)
```
Comment (amasend) [inline simulation/container.py]: The question is, do we preserve individual names/ids that way?
Eg. during visualization we would want to show particular individuals, after such removal and new object creation there will be weird consequences probably. We should know after some visualization implementation.

**Review by amasend — COMMENTED** [review]
I've tested this implementation with 3k population.
Computation times were ~8 seconds :( even when period between steps was set to 1 second.
For sure this implementation is not good for such simulation but when we consider simulation up to hmm.. ~1k population it might be reasonable. This needs more tests.

The performance issue is probably related with objects creation and distance comparison. We are comparing one to all positions what is really not efficient O(N^2).
Maybe it is worth to move all individuals at the same time, then check distance and eliminate each current object from next check? I guess it will be O(N).

**Comment by amasend** [comment]
I was thinking about performance upgrade.
Consider:
 - one common container with all individuals
 - separate container holding S individuals or S locations
 - separate container holding I individuals or I locations

Movements are computed once per global iteration (every individual from global container will change place). Second stage could be iterate over every S individual in S container and check if we are in infection range from any individual from I container. (Here we can use some kind of sorting algorithm, each of S and I container could be sorted based on `x` and `y` values, in that way we can faster match infection distances right?). Third stage could be iteration over I container and checking recover probabilities.
Removing and adding individuals from and to S and I containers could be done after we store indexes of that individuals and after full stage loops (to mitigate some unwanted list behavior).

**Comment by Zavioer** [comment]
> I was thinking about performance upgrade.
> Consider:

Ok, I try to implement this solution. Now I implemented solution with one class for S, I, R individuals and I think is a little bit better but when interval time is below 1 second it is not working correctly. I do not change logic, so i commit what I have done and I start to implement solution above.

**Comment by amasend** [comment]
@Zavioer I am just saying one possible solution, you do not need to implement that ;) just let me check the new commited code

File: simulation/container_interface.py [hunk 6]
```diff
@@ -49,19 +48,17 @@ def is_alive(self) -> bool:
...
+                      infection_probability: float, recover_probability: float,
+                      dead_probability: float, infection_range: float) -> None:
+        for i in range(instance_amount):
+            x = random.uniform(0, self.width - 1)
+            y = random.uniform(0, self.height - 1)
+            self.person = groups.Person(x=0, y=0, infection_probability=infection_probability,
+                                        recover_probability=recover_probability,
+                                        dead_probability=dead_probability, infection_range=infection_range)
+            self.person.x = x
+            self.person.y = y
```
Comment (amasend) [inline simulation/container_interface.py:61]: Why `x` and `y` are not set in `Person` constructor to these values? Is there some reason?

File: simulation/groups_interface.py [hunk 7]
```diff
@@ -16,9 +16,16 @@ class PopulationGroup(metaclass=ABCMeta):
     id = count(0)
```
Comment (amasend) [inline simulation/groups_interface.py:16]: Do we still need this `id`? If not, maybe it is better to clean up these things.

**Review by amasend — APPROVED** [review]
Ok! This implementation can handle ~3k population with ~1-1.5s intervals ;) nice, but could be better! Python can be faster, much more faster :dagger:
For the sake of knowledge, you can try measure times for appending a list with two methods:
 - 
```python
[some_list.append(1) for i in range(100000)]
```
 - 
```python
append_me = some_list.append
[append_me(1) for i in range(100000]
```
But why and how? What could be the benefit and drawback?

**Pull request merged**